File: scripts/node/parsers/externalCredentialParser.js
'use strict';

const fs   = require('fs');
const path = require('path');

function extractTag(xml, tag) {
  const m = xml.match(new RegExp(`<${tag}[^>]*>([\\s\\S]*?)<\\/${tag}>`));
  return m ? m[1].trim() : null;
}

/**
 * Reads External Credential metadata from SFDX source format.
 * @param {string} typeDir - path to the externalCredentials/ folder
 * @returns {{ credentials: Array }}
 */
async function parseExternalCredentials(typeDir) {
  const credentials = [];
  if (!fs.existsSync(typeDir)) return { credentials };

  const files = fs.readdirSync(typeDir).filter(f => f.endsWith('.externalCredential-meta.xml'));

  for (const file of files) {
    const xml = fs.readFileSync(path.join(typeDir, file), 'utf8');
    credentials.push({
      apiName:                file.replace('.externalCredential-meta.xml', ''),
      label:                  extractTag(xml, 'label'),
      description:            extractTag(xml, 'description'),
      authenticationProtocol: extractTag(xml, 'authenticationProtocol'),
    });
  }

  return { credentials };
}

module.exports = { parseExternalCredentials };

File: scripts/python/parsers/external_credential_parser.py
"""Parser for External Credential metadata (.externalCredential-meta.xml)."""
import re
from pathlib import Path


def _extract_tag(xml: str, tag: str) -> str | None:
    m = re.search(rf'<{tag}[^>]*>([\s\S]*?)</{tag}>', xml)
    return m.group(1).strip() if m else None


def parse_external_credentials(type_dir: str) -> dict:
    """
    @param type_dir: path to the externalCredentials/ folder
    @returns: { credentials: list }
    """
    credentials = []
    p = Path(type_dir)
    if not p.exists():
        return {'credentials': credentials}

    for f in sorted(p.glob('*.externalCredential-meta.xml')):
        xml = f.read_text(encoding='utf-8')
        credentials.append({
            'apiName':                f.name.replace('.externalCredential-meta.xml', ''),
            'label':                  _extract_tag(xml, 'label'),
            'description':            _extract_tag(xml, 'description'),
            'authenticationProtocol': _extract_tag(xml, 'authenticationProtocol'),
        })

    return {'credentials': credentials}
